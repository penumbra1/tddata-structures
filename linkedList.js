export function createNode(value) {
  return {
    value,
    next: null
  };
}

// Singly linked list
export function createLinkedList() {
  return {
    head: null,
    tail: null,
    length: 0,
    push(value) {
      const node = createNode(value);

      if (this.length === 0) {
        this.head = node;
      } else {
        this.tail.next = node;
      }

      this.tail = node;
      this.length++;
    },
    pop() {
      if (this.length === 0) {
        return undefined;
      }

      const oldTail = this.tail;
      // newTail will refer to the node previous to oldTail
      let newTail = this.head;

      if (this.length === 1) {
        this.head = null;
        newTail = null;
      } else {
        while (newTail.next !== this.tail) {
          newTail = newTail.next;
        }

        newTail.next = null;
      }

      this.tail = newTail;
      this.length--;

      return oldTail;
    },
    shift(value) {
      const node = createNode(value);

      node.next = this.head;
      this.head = node;
      if (this.length === 0) {
        this.tail = node;
      }
      this.length++;
    },
    unshift() {
      if (this.length === 0) {
        return undefined;
      }

      if (this.length === 1) {
        this.tail = null;
      }

      const oldHead = this.head;
      this.head = this.head.next;
      this.length--;

      return oldHead;
    },
    insertAfter(prevValue, value) {
      const node = createNode(value);

      let current = this.head;

      while (current !== null && current.value !== prevValue) {
        current = current.next;
      }

      if (current === null) {
        throw new Error("Previous node not found");
      }

      if (current === this.tail) {
        this.tail = node;
      }

      node.next = current.next;
      current.next = node;
      this.length++;
    },
    get isEmpty() {
      return this.length === 0;
    }
  };
}
