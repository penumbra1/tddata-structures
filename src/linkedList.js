export function createNode(value) {
  return {
    value,
    next: null
  };
}

// Singly linked list
export function createLinkedList() {
  return {
    head: null,
    tail: null,
    length: 0,
    get isEmpty() {
      return this.length === 0;
    },
    get(index) {
      if (index < 0 || index > this.length - 1) return null;

      if (index === 0) return this.head;
      if (index === this.length - 1) return this.tail;

      let current = this.head;

      let i = 0;
      while (i !== index) {
        current = current.next;
        i++;
      }

      return current;
    },
    delete(index) {
      if (index < 0 || index > this.length - 1) return null;

      let currentIndex = 0;
      let [previous, current, following] = [null, this.head, this.head.next];

      while (currentIndex < index) {
        [previous, current, following] = [current, following, following.next];
        currentIndex++;
      }

      if (previous === null) {
        // Current head will be removed
        this.head = following;
      } else {
        previous.next = following;
      }

      if (following === null) {
        // Current tail will be removed
        this.tail = previous;
      }

      this.length--;

      return current;
    },
    push(value) {
      const node = createNode(value);

      if (this.length === 0) {
        this.head = node;
      } else {
        this.tail.next = node;
      }

      this.tail = node;
      this.length++;
    },
    pop() {
      if (this.length === 0) {
        return null;
      }

      const oldTail = this.tail;
      // newTail will refer to the node previous to oldTail
      let newTail = this.head;

      if (this.length === 1) {
        this.head = null;
        newTail = null;
      } else {
        while (newTail.next !== this.tail) {
          newTail = newTail.next;
        }

        newTail.next = null;
      }

      this.tail = newTail;
      this.length--;

      return oldTail;
    },
    shift(value) {
      const node = createNode(value);

      node.next = this.head;
      this.head = node;
      if (this.length === 0) {
        this.tail = node;
      }
      this.length++;
    },
    unshift() {
      if (this.length === 0) {
        return null;
      }

      if (this.length === 1) {
        this.tail = null;
      }

      const oldHead = this.head;
      this.head = this.head.next;
      this.length--;

      return oldHead;
    },
    insertAt(index, value) {
      const node = createNode(value);

      if (index < 0 || index > this.length) return null;

      let currentIndex = 0;
      let [insertAfter, insertBefore] = [null, this.head];

      if (index === this.length) {
        // No need to traverse the list - fast forward to the end:
        currentIndex = this.length;
        [insertAfter, insertBefore] = [this.tail, null];
        this.tail = node;
      }

      while (currentIndex < index) {
        [insertAfter, insertBefore] = [insertBefore, insertBefore.next];
        currentIndex++;
      }

      if (insertAfter === null) {
        this.head = node;
      } else {
        insertAfter.next = node;
      }

      node.next = insertBefore;

      this.length++;

      return node;
    },
    find(value) {
      let current = this.head;
      let index = 0;

      while (current && current.value !== value) {
        current = current.next;
        index++;
      }

      if (current === null) return null;

      return index;
    },
    print() {
      let current = this.head;
      const values = [];

      while (current) {
        values.push(current.value);
        current = current.next;
      }

      return values.join(" => ");
    }
  };
}
