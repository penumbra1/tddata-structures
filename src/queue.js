export function createQueue() {
  const queue = [];

  return {
    enqueue(item) {
      queue.unshift(item);
    },
    dequeue() {
      return queue.pop();
    },
    peek() {
      return queue[queue.length - 1];
    },
    get length() {
      return queue.length;
    },
    get isEmpty() {
      return queue.length === 0;
    }
  };
}

export function createPriorityQueue() {
  const lowPriority = createQueue();
  const highPriority = createQueue();
  return {
    enqueue(item, isHighPriority = false) {
      if (isHighPriority) highPriority.enqueue(item);
      else lowPriority.enqueue(item);
    },
    dequeue() {
      if (!highPriority.isEmpty) return highPriority.dequeue();
      return lowPriority.dequeue();
    },
    peek() {
      if (!highPriority.isEmpty) return highPriority.peek();
      return lowPriority.peek();
    },
    get length() {
      return highPriority.length + lowPriority.length;
    },
    get isEmpty() {
      return highPriority.isEmpty && lowPriority.isEmpty;
    }
  };
}
