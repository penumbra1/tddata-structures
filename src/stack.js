export default function createStack(max = 5) {
  const stack = [];

  return {
    max,
    push(item) {
      if (stack.length < max) stack.push(item);
      else throw new Error("Stack overflow!");
    },
    pop() {
      return stack.pop();
    },
    peek() {
      return stack[stack.length - 1];
    },
    get isEmpty() {
      return stack.length === 0;
    },
    get length() {
      return stack.length;
    }
  };
}
