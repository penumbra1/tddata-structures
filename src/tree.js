export function createNode(key) {
  const children = [];

  return {
    key,
    children,
    addChild(childKey) {
      const child = createNode(childKey);
      children.push(child);
      return child;
    }
  };
}

export function createTree(rootKey) {
  if (rootKey == null) throw new Error("Creating a tree requires a root key");

  const root = createNode(rootKey);

  return {
    root,
    traverse(node, action, depth) {
      action(node, depth);

      if (node.children.length) {
        node.children.forEach(child => {
          this.traverse(child, action, depth + 1);
        });
      }
    },
    print() {
      let result = "";
      const addLine = (node, depth) => {
        result += depth
          ? `${"|  ".repeat(depth)}|--${node.key}\n`
          : `|--${node.key}\n`;
      };

      this.traverse(this.root, addLine, 0);

      return result;
    }
  };
}
