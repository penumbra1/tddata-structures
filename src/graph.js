import { createQueue } from "./queue";

export function createNode(key) {
  const neighbors = [];

  return {
    key,
    neighbors,
    addNeighbor(node) {
      neighbors.push(node);
    },
    print() {
      const neighborList = neighbors.map(n => n.key).join(" ");
      const line = `${this.key}${neighborList && ` => ${neighborList}`}`;

      return line;
    }
  };
}

export function createGraph(directed = false) {
  const nodes = [];
  const edges = [];

  return {
    directed,
    nodes,
    edges,
    getNode(key) {
      return nodes.find(node => node.key === key) || null;
    },
    addNode(key) {
      if (key == null || this.getNode(key) != null) return null;

      const newNode = createNode(key);
      nodes.push(newNode);

      return newNode;
    },
    addEdge(node1Key, node2Key) {
      const [node1, node2] = [this.getNode(node1Key), this.getNode(node2Key)];

      if (!node1 || !node2) return null;

      node1.addNeighbor(node2);
      if (!this.directed) node2.addNeighbor(node1);

      const newEdge = `${node1Key}-${node2Key}`;

      edges.push(newEdge);

      return newEdge;
    },
    print() {
      return nodes.map(node => node.print()).join("\n");
    },
    traverseBreadth(startingKey, action) {
      const startingNode = this.getNode(startingKey);
      const visited = {};
      let currentNode;
      const queue = createQueue();
      if (startingNode) queue.enqueue(startingNode);
      else throw new Error(`Starting node not found`);

      while (!queue.isEmpty) {
        currentNode = queue.dequeue();
        currentNode.neighbors.forEach(node => {
          if (!visited[node.key]) queue.enqueue(node);
        });
        action(currentNode);
        visited[currentNode.key] = true;
      }
    },
    traverseDepth(startingKey, action) {
      const startingNode = this.getNode(startingKey);
      const visited = {};

      function explore(node) {
        if (visited[node.key]) return;

        action(node);
        visited[node.key] = true;
        node.neighbors.forEach(neighbor => explore(neighbor));
      }

      if (startingNode) explore(startingNode);
      else throw new Error(`Starting node not found`);
    }
  };
}
