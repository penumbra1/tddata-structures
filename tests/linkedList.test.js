import { createNode, createLinkedList } from "../src/linkedList";

describe("Node", () => {
  test("node factory should create a node", () => {
    const n = createNode(1);

    expect(n).toEqual({ value: 1, next: null });
  });
});

describe("LinkedList", () => {
  let l;

  beforeEach(() => {
    l = createLinkedList();
  });

  test("linkedList factory should create an empty list", () => {
    expect(l.head).toBe(null);
    expect(l.tail).toBe(null);
    expect(l.length).toBe(0);
    expect(l.isEmpty).toBe(true);
  });

  test("push should add a new tail (and head if the list was empty)", () => {
    const [one, two] = [createNode(1), createNode(2)];

    l.push(1);

    expect(l.length).toBe(1);
    expect(l.head).toEqual(one);
    expect(l.tail).toEqual(one);

    l.push(2);

    expect(l.length).toBe(2);
    expect(l.head).toEqual({ value: 1, next: two });
    expect(l.tail).toEqual(two);
  });

  test("pop should remove and reset the tail", () => {
    const [two, three] = [createNode(2), createNode(3)];

    l.push(1);
    l.push(2);
    l.push(3);
    const removed = l.pop();

    expect(removed).toEqual(three);
    expect(l.tail).toEqual(two);
    expect(l.head).toEqual({ value: 1, next: two });
    expect(l.length).toBe(2);
  });

  test("if pop leaves a single node, it should be both head and tail", () => {
    const one = createNode(1);

    l.push(1);
    l.push(2);
    l.pop();

    expect(l.tail).toEqual(one);
    expect(l.head).toEqual(one);
    expect(l.length).toBe(1);
  });

  test("pop should return null if the list is empty", () => {
    expect(l.pop()).toBe(null);
  });

  test("pop should result in an empty list if length was 1", () => {
    l.push(1);
    l.pop();

    expect(l.head).toBe(null);
    expect(l.tail).toBe(null);
    expect(l.length).toBe(0);
    expect(l.isEmpty).toBe(true);
  });

  test("insertAt should insert a node into an empty list", () => {
    const one = createNode(1);

    const node = l.insertAt(0, 1);

    expect(node).toEqual(one);
    expect(l.head).toEqual(one);
    expect(l.tail).toEqual(one);
    expect(l.length).toBe(1);
  });

  test("insertAt should insert a node into a non-empty list", () => {
    l.insertAt(0, 1);
    const tailNode = l.insertAt(1, 3);

    expect(tailNode).toEqual(createNode(3));
    expect(l.head).toEqual({ value: 1, next: tailNode });
    expect(l.tail).toEqual(tailNode);
    expect(l.length).toBe(2);

    const middleNode = l.insertAt(1, 2);

    expect(middleNode).toEqual({ value: 2, next: tailNode });
    expect(l.head).toEqual({ value: 1, next: middleNode });
    expect(l.tail).toEqual(tailNode);
    expect(l.length).toBe(3);
  });

  test("insertAt should do nothing and return null if index is out of range", () => {
    expect(l.insertAt(-1, 1)).toBe(null);
    expect(l.insertAt(1, 1)).toBe(null);
    expect(l.length).toBe(0);

    l.push(1);
    l.push(1);

    expect(l.insertAt(3, 1)).toBe(null);
    expect(l.length).toBe(2);
  });

  test("shift should insert a new head (and tail if the list was empty)", () => {
    const one = createNode(1);

    l.shift(1);

    expect(l.head).toEqual(one);
    expect(l.tail).toEqual(one);
    expect(l.length).toBe(1);

    l.shift(2);

    expect(l.head).toEqual({ value: 2, next: one });
    expect(l.tail).toEqual(one);
    expect(l.length).toBe(2);
  });

  test("unshift should remove and reset the head", () => {
    const one = createNode(1);

    l.shift(1);
    l.shift(2);
    l.shift(3);
    const removed = l.unshift();
    const two = { value: 2, next: one };

    expect(removed).toEqual({ value: 3, next: two });
    expect(l.head).toEqual(two);
    expect(l.tail).toEqual(one);
    expect(l.length).toBe(2);
  });

  test("if unshift leaves a single node, it should be both head and tail", () => {
    const one = createNode(1);

    l.shift(1);
    l.shift(2);
    l.unshift();

    expect(l.tail).toEqual(one);
    expect(l.head).toEqual(one);
    expect(l.length).toBe(1);
  });

  test("unshift should result in an empty list if length was 1", () => {
    l.shift(1);
    const removed = l.unshift();

    expect(removed).toEqual({ value: 1, next: null });
    expect(l.head).toBe(null);
    expect(l.tail).toBe(null);
    expect(l.length).toBe(0);
    expect(l.isEmpty).toBe(true);
  });

  test("unshift should return null if the list was empty", () => {
    expect(l.unshift()).toBe(null);
  });

  test("find should return null if list is empty", () => {
    expect(l.find(1)).toBe(null);
  });

  test("find should return the correct index", () => {
    l.push(1);

    expect(l.find(1)).toBe(0);

    l.push(2);
    l.push(3);

    expect(l.find(2)).toBe(1);
    expect(l.find(3)).toBe(2);
  });

  test("find should return the index of the first occurrence", () => {
    l.push(1);
    l.push(2);
    l.push(2);

    expect(l.find(2)).toBe(1);
  });

  test("get should return null if index is out of range", () => {
    expect(l.get(0)).toBe(null);
    expect(l.get(-1)).toBe(null);

    l.unshift(1);
    expect(l.get(2)).toBe(null);
  });

  test("get should return the correct node", () => {
    const [one, three] = [createNode(1), createNode(3)];

    l.push(1);

    expect(l.get(0)).toEqual(one);

    l.push(2);
    l.push(3);

    expect(l.get(1)).toEqual({ value: 2, next: three });
    expect(l.get(2)).toEqual(three);
  });

  test("delete should do nothing and return null if index is out of range", () => {
    expect(l.delete(-1)).toBe(null);
    expect(l.delete(0)).toBe(null);
    expect(l.length).toBe(0);

    l.push(1);

    expect(l.delete(1)).toBe(null);
    expect(l.length).toBe(1);
  });

  test("delete should return the deleted node", () => {
    l.push(1);
    const deleted = l.delete(0);

    expect(deleted).toEqual(createNode(1));
  });

  test("delete should result in an empty list if length was 1", () => {
    l.push(1);

    l.delete(0);

    expect(l.head).toBe(null);
    expect(l.tail).toBe(null);
    expect(l.length).toBe(0);
    expect(l.isEmpty).toBe(true);
  });

  test("delete should remove and reset head or tail", () => {
    const [one, two] = [createNode(1), createNode(2)];
    const [headIndex, tailIndex] = [0, 1];

    l.push(1);
    l.push(2); // [1, 2]
    l.delete(tailIndex);

    expect(l.head).toEqual(one);
    expect(l.tail).toEqual(one);
    expect(l.length).toBe(1);

    l.push(1);
    l.push(2); // [1, 1, 2]
    l.delete(headIndex);

    expect(l.head).toEqual({ value: 1, next: two });
    expect(l.tail).toEqual(two);
    expect(l.length).toBe(2);
  });

  test("print should return the correct string representation", () => {
    expect(l.print()).toBe("");

    l.push(1);

    expect(l.print()).toBe("1");

    l.push(2);
    l.push(3);
    l.push(4);

    expect(l.print()).toMatchSnapshot();
  });
});
