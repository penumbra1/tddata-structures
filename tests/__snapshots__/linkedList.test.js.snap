// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linkedList print should return the correct string representation 1`] = `"1 => 2 => 3 => 4"`;
